datafusion/core/src/physical_plan/joins/nested_loop_join.rs

```diff
@@ -0,0 +1,872 @@
+                            if need_produce_result_in_final(self.join_type) {
+                                left_side.iter().flatten().for_each(|x| {
+                                    visited_left_side.set_bit(x as usize, true);
+                                });
+                            }
+                            // adjust the two side indices base on the join type
+                            let (left_side, right_side) = adjust_indices_by_join_type(
+                                left_side,
+                                right_side,
+                                right_batch.num_rows(),
+                                self.join_type,
+                            );
+
+                            let result = build_batch_from_indices(
+                                &self.schema,
+                                left_data,
+                                &right_batch,
+                                left_side,
+                                right_side,
+                                &self.column_indices,
+                            );
+                            Some(result)
+                        }
+                        Err(e) => Some(Err(e)),
+                    };
+                    result
+                }
+                Some(err) => Some(err),
+                None => {
+                    if need_produce_result_in_final(self.join_type) && !self.is_exhausted
+                    {
+                        // use the global left bitmap to produce the left indices and right indices
+                        let (left_side, right_side) =
+                            get_final_indices(visited_left_side, self.join_type);
+                        let empty_right_batch =
+                            RecordBatch::new_empty(self.right.schema());
+                        // use the left and right indices to produce the batch result
+                        let result = build_batch_from_indices(
+                            &self.schema,
+                            left_data,
+                            &empty_right_batch,
+                            left_side,
+                            right_side,
+                            &self.column_indices,
+                        );
+                        self.is_exhausted = true;
+                        Some(result)
+                    } else {
+                        // end of the join loop
+                        None
+                    }
+                }
+            })
+    }
+}
+
+impl Stream for NestedLoopJoinStream {
+    type Item = ArrowResult<RecordBatch>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        self.poll_next_impl(cx)
+    }
+}
+
+impl RecordBatchStream for NestedLoopJoinStream {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
there are som duplicated test code in the `hashjoin` and `cross join`.

I will refactor and clean up them in the followup pr